### PR TITLE
[Added] the new feature to ui-grid for multi-select filter

### PR DESCRIPTION
**Pull request by sanjaynishad**
Added the **new feature** to ui-grid for **multi-select  filter**, it's inspired from the example http://ui-grid.info/docs/#/tutorial/306_custom_filters

already added new repo--> https://github.com/sanjaynishad/ui-grid-multiselect-filter
demo -->http://www.sanjaynishad.com/ui-grid-multiselect-filter/
Thanks..:)

### CODE REVIEW

**Comment by dlgski** [comment]
@sanjaynishad can you please fix this PR so it passes integration?  Thank you

**Comment by sanjaynishad** [comment]
Sure, please give me some time. Thanks

**Comment by sanjaynishad** [comment]
@dlgski fixed the PR, Thank You.. :)

File: src/less/multiselect-filter.less [hunk 1]
```diff
@@ -0,0 +1,33 @@
...
+
+.ui-grid-multi-select-filter .header {
+	font-size: 15px;
+	font-weight: bold;
+	margin-bottom: 5px;
+	text-transform: capitalize;
+}
+
+.ui-grid-multi-select-filter .footer {
+
```
Comment (dlgski) [inline src/less/multiselect-filter.less]: remove ?

File: src/js/core/directives/ui-grid-multiselect-filter.js [hunk 2]
```diff
@@ -0,0 +1,99 @@
+if (!Element.prototype.matches) {
```
Comment (dlgski) [inline src/js/core/directives/ui-grid-multiselect-filter.js]: this should be contained w/in an closure.   Also needs 'use strict'

Comment (dlgski) [inline src/js/core/directives/ui-grid-multiselect-filter.js]: why do you have all this code outside of the module?

File: src/js/core/directives/ui-grid-multiselect-filter.js [hunk 3]
```diff
@@ -0,0 +1,99 @@
+if (!Element.prototype.matches) {
+  Element.prototype.matches = Element.prototype.msMatchesSelector;
+}
+if (!Element.prototype.closest) {
```
Comment (dlgski) [inline src/js/core/directives/ui-grid-multiselect-filter.js]: have you looked at gridUtil ?

Comment (sanjaynishad) [inline src/js/core/directives/ui-grid-multiselect-filter.js]: I have added these code for IE, btw no need to keep this. removed the same

File: src/less/multiselect-filter.less (hunk 1 above)
Comment (sanjaynishad) [inline src/less/multiselect-filter.less]: ?

File: src/js/core/directives/ui-grid-multiselect-filter.js (hunk 2 above)
Comment (sanjaynishad) [inline src/js/core/directives/ui-grid-multiselect-filter.js]: done

File: src/less/multiselect-filter.less (hunk 1 above)
Comment (sanjaynishad) [inline src/less/multiselect-filter.less]: done

**Comment by franbenz** [comment]
Hi @dlgski, any tips on next steps for this PR?

**Comment by mportuga** [comment]
@sanjaynishad and @franbenz Can you please add unit tests to your feature?

**Comment by mportuga** [comment]
If I see no response in 4 days or so, I will be declining this PR for now.

**Comment by sanjaynishad** [comment]
@mportuga I'm not getting time to work on unit test.
can someone please create? otherwise I'll do once I'll free.

**Comment by mportuga** [comment]
I might try if I get some time, but for now, this will just have to stay open.

**Comment by pradapkumar** [comment]
Hi Sanjay
 I tried adding your multi-select filter and when I click "..." button below the column name first time its shows me the pop up and allow me to filter and the grid gets filtered but when I click the same button (...) again second time I getting error and the pop up is not coming up

below is the error 

Error: [$compile:ctreq] http://errors.angularjs.org/1.4.8/$compile/ctreq?p0=uiGrid&p1=uiGridHeaderCell
   at p (http://ff01faf5b908e.corp.bankofamerica.com:8080/Scripts/angular.min.js:59:407)
   at p (http://ff01faf5b908e.corp.bankofamerica.com:8080/Scripts/angular.min.js:59:475)
   at D (http://ff01faf5b908e.corp.bankofamerica.com:8080/Scripts/angular.min.js:62:1)
   at g (http://ff01faf5b908e.corp.bankofamerica.com:8080/Scripts/angular.min.js:55:105)
   at Anonymous function (http://ff01faf5b908e.corp.bankofamerica.com:8080/Scripts/angular.min.js:54:249)
   at Anonymous function (http://ff01faf5b908e.corp.bankofamerica.com:8080/Scripts/angular.min.js:56:72)
   at k (http://ff01faf5b908e.corp.bankofamerica.com:8080/Scripts/angular.min.js:60:370)
   at Anonymous function (http://ff01faf5b908e.corp.bankofamerica.com:8080/Scripts/angular.min.js:281:253)
   at Anonymous function (http://ff01faf5b908e.corp.bankofamerica.com:8080/Scripts/angular.min.js:130:77)
   at r.prototype.$digest (http://ff01faf5b908e.corp.bankofamerica.com:8080/Scripts/angular.min.js:131:147)

I have added the gridoption

$scope.gridOptions = {
        enableSorting: true,
        paginationPageSizes: [25, 50, 75],
        paginationPageSize: 25,
        enablePaginationControls: false,
        enableGridMenu: false,
        gridMenuShowHideColumns: false,
        exporterMenuPdf: false,
        enableRowHeaderSelection: false,
        enableHorizontalScrollbar: uiGridConstants.scrollbars.NEVER,
        enableRowSelection: true,       
        enableColumnMenus: false,
        enableFiltering: true,
        columnDefs: $scope.columns,
        showGridFooter: true,
      
    };
and header template as below.
field: fieldName, displayName: fieldName, width: currentFieldWidth,
                        filterHeaderTemplate: '<div multi-select-filter on-filter="grid.appScope.filtered($terms)"></div>', enableCellEdit: grdFieldisEditable

Please let me know what mistake I have done.

Thanks

**Comment by sanjaynishad** [comment]
Hi @pradapkumar ,
Sorry, I'm not able to reproduce this issue, it works fine for me. I'll check again one get time.
Thanks.

**Comment by dlgski** [comment]
I cannot merge feature updates without unit tests.  Is this PR still necessary?

**Comment by sanjaynishad** [comment]
Will add unit test

**Comment by mportuga** [comment]
@sanjaynishad Any progress on unit tests?

**Comment by sanjaynishad** [comment]
@mportuga I didn't get time.
@dlgski if no one is going to add unit test, you may close this PR

**Comment by mportuga** [comment]
Closing due to lack of tests.

**Comment by pradapkumar** [comment]
@sanjaynishad no prb thanks 😊

**Pull request closed without merging**